File: Code Matlab/07. Physical Models ROMs/07.1 POD on Fluid Flow past a Cylinder/CylinderFlowModes.m
clc, clear, close all
%% Import Data
[gif,map] = imread('CylinderFlow','gif','Frames','all');

%% Resample

for i = 1:226
    img = double(gif(:,:,:,i));
    img = img(91:180,51:end);
    X(:,i) = reshape(img,90*430,1);
    gif2(:,:,:,i) = mat2gray(img); 
end

implay(gif2);

%% Principal Modes
% PCA of the timeframes

[U,S,V] = svd(X,'econ');
S = diag(S);

figure('Name','Sing Values')
subplot(1,2,1)
semilogy(S(1:50),'o-')
hold on
title('Singular Values')
xlim([1,50])
subplot(1,2,2)
for i = 1:length(S)
    svar(i) = sum(S(1:i))/sum(S);
end
plot(svar(1:226));
xlim([1,226]);
title('Captured Variance')

%% Plot first PODs

figure('Name','Principal Flow Modes')
subplot('Position',[0.34,0.77,0.35,0.2])
imshow(mat2gray(reshape(abs(U(:,1)),90,430))), colormap('jet');
title('Mean Flow (POD 1)')
subplot(5,2,3)
imshow(mat2gray(reshape(abs(U(:,2)),90,430))), colormap('jet');
title('POD 2')
subplot(5,2,4)
imshow(mat2gray(reshape(abs(U(:,3)),90,430))), colormap('jet');
title('POD 3')
subplot(5,2,5)
imshow(mat2gray(reshape(abs(U(:,4)),90,430))), colormap('jet');
title('POD 4')
subplot(5,2,6)
imshow(mat2gray(reshape(abs(U(:,5)),90,430))), colormap('jet');
title('POD 5')
subplot(5,2,7)
imshow(mat2gray(reshape(abs(U(:,6)),90,430))), colormap('jet');
title('POD 6')
subplot(5,2,8)
imshow(mat2gray(reshape(abs(U(:,7)),90,430))), colormap('jet');
title('POD 7')
subplot(5,2,9)
imshow(mat2gray(reshape(abs(U(:,8)),90,430))), colormap('jet');
title('POD 8')
subplot(5,2,10)
imshow(mat2gray(reshape(abs(U(:,9)),90,430))), colormap('jet');
title('POD 9')

%% Time Correlation


figure('Name','Principal Flow Modes')
subplot(9,2,1)
imshow(mat2gray(reshape(abs(U(:,1)),90,430))), colormap('jet');
title('POD Modes')
subplot(9,2,2)
title('Corellation to Time')
hold on
plot(V(:,1));
xlim([1,226])
subplot(9,2,3)
imshow(mat2gray(reshape(abs(U(:,2)),90,430))), colormap('jet');
subplot(9,2,4)
plot(V(:,2));
xlim([1,226])
subplot(9,2,5)
imshow(mat2gray(reshape(abs(U(:,3)),90,430))), colormap('jet');
subplot(9,2,6)
plot(V(:,3));
xlim([1,226])
subplot(9,2,7)
imshow(mat2gray(reshape(abs(U(:,4)),90,430))), colormap('jet');
subplot(9,2,8)
plot(V(:,4));
xlim([1,226])
subplot(9,2,9)
imshow(mat2gray(reshape(abs(U(:,5)),90,430))), colormap('jet');
subplot(9,2,10)
plot(V(:,5));
xlim([1,226])
subplot(9,2,11)
imshow(mat2gray(reshape(abs(U(:,6)),90,430))), colormap('jet');
subplot(9,2,12)
plot(V(:,6));
xlim([1,226])
subplot(9,2,13)
imshow(mat2gray(reshape(abs(U(:,7)),90,430))), colormap('jet');
subplot(9,2,14)
plot(V(:,7));
xlim([1,226])
subplot(9,2,15)
imshow(mat2gray(reshape(abs(U(:,8)),90,430))), colormap('jet');
subplot(9,2,16)
plot(V(:,8));
xlim([1,226])
subplot(9,2,17)
imshow(mat2gray(reshape(abs(U(:,9)),90,430))), colormap('jet');
subplot(9,2,18)
plot(V(:,9));
xlim([1,226])

%% Compressed Sensing with time samples
% Only 22 randomly taken time samples

%R = round(rand(1,30)*226);

% Optimisation
R = round(linspace(1,216,40));
R = R+randi(2,1,40);

Xr = X(:,R);

Xa = zeros(size(X)); Xa(:,R)=X(:,R); 
for i = 1:226
    img = reshape(double(Xa(:,i)),90,430);
    gif2(:,:,:,i) = mat2gray(img); 
end
implay(gif2);

[Ur,Sr,Vr] = svd(Xr,'econ');

figure('Name','Principal Flow Modes')
subplot('Position',[0.34,0.77,0.35,0.2])
imshow(mat2gray(reshape(abs(Ur(:,1)),90,430))), colormap('jet');
title('Mean flow (POD 1)')
subplot(5,2,3)
imshow(mat2gray(reshape(abs(Ur(:,2)),90,430))), colormap('jet');
title('POD 2')
subplot(5,2,4)
imshow(mat2gray(reshape(abs(Ur(:,3)),90,430))), colormap('jet');
title('POD 3')
subplot(5,2,5)
imshow(mat2gray(reshape(abs(Ur(:,4)),90,430))), colormap('jet');
title('POD 4')
subplot(5,2,6)
imshow(mat2gray(reshape(abs(Ur(:,5)),90,430))), colormap('jet');
title('POD 5')
subplot(5,2,7)
imshow(mat2gray(reshape(abs(Ur(:,6)),90,430))), colormap('jet');
title('POD 6')
subplot(5,2,8)
imshow(mat2gray(reshape(abs(Ur(:,7)),90,430))), colormap('jet');
title('POD 7')
subplot(5,2,9)
imshow(mat2gray(reshape(abs(Ur(:,8)),90,430))), colormap('jet');
title('POD 8')
subplot(5,2,10)
imshow(mat2gray(reshape(abs(Ur(:,9)),90,430))), colormap('jet');
title('POD 9')

Phi = conj(dftmtx(226)/226);
PSI = Phi(R,:);

%%

for i=1:30
b = Vr(:,i);       
cvx_begin;
    variable xL1(size(PSI,2)) complex; % sparse vector of coefficients 
    minimize( norm(xL1,1) );
    subject to
        norm(PSI*xL1 - b,2) <= 0.005;
cvx_end;
Vrec(:,i) = real(ifft(xL1));  
end

%% Visualize downsampled PODs

figure('Name','Principal Flow Modes downsampled')
subplot(9,2,1)
imshow(mat2gray(reshape(abs(Ur(:,1)),90,430))), colormap('jet');
title('POD Modes downsampled')
subplot(9,2,2)
title('Corellation to Time downsampled')
hold on
plot(Vr(:,1));
xlim([1,30])
subplot(9,2,3)
imshow(mat2gray(reshape(abs(Ur(:,2)),90,430))), colormap('jet');
subplot(9,2,4)
plot(Vr(:,2));
xlim([1,30])
subplot(9,2,5)
imshow(mat2gray(reshape(abs(Ur(:,3)),90,430))), colormap('jet');
subplot(9,2,6)
plot(Vr(:,3));
xlim([1,30])
subplot(9,2,7)
imshow(mat2gray(reshape(abs(Ur(:,4)),90,430))), colormap('jet');
subplot(9,2,8)
plot(Vr(:,4));
xlim([1,30])
subplot(9,2,9)
imshow(mat2gray(reshape(abs(Ur(:,5)),90,430))), colormap('jet');
subplot(9,2,10)
plot(Vr(:,5));
xlim([1,30])
subplot(9,2,11)
imshow(mat2gray(reshape(abs(Ur(:,6)),90,430))), colormap('jet');
subplot(9,2,12)
plot(Vr(:,6));
xlim([1,30])
subplot(9,2,13)
imshow(mat2gray(reshape(abs(Ur(:,7)),90,430))), colormap('jet');
subplot(9,2,14)
plot(Vr(:,7));
xlim([1,30])
subplot(9,2,15)
imshow(mat2gray(reshape(abs(Ur(:,8)),90,430))), colormap('jet');
subplot(9,2,16)
plot(Vr(:,8));
xlim([1,30])
subplot(9,2,17)
imshow(mat2gray(reshape(abs(Ur(:,9)),90,430))), colormap('jet');
subplot(9,2,18)
plot(Vr(:,9));
xlim([1,30])


figure('Name','Principal Flow Modes downsampled')
subplot(9,2,1)
imshow(mat2gray(reshape(abs(Ur(:,1)),90,430))), colormap('jet');
title('POD Modes downsampled')
subplot(9,2,2)
title('Corellation to Time reconstructed')
hold on
plot(Vrec(:,1));
xlim([1,226])
subplot(9,2,3)
imshow(mat2gray(reshape(abs(Ur(:,2)),90,430))), colormap('jet');
subplot(9,2,4)
plot(Vrec(:,2));
xlim([1,226])
subplot(9,2,5)
imshow(mat2gray(reshape(abs(Ur(:,3)),90,430))), colormap('jet');
subplot(9,2,6)
plot(Vrec(:,3));
xlim([1,226])
subplot(9,2,7)
imshow(mat2gray(reshape(abs(Ur(:,4)),90,430))), colormap('jet');
subplot(9,2,8)
plot(Vrec(:,4));
xlim([1,226])
subplot(9,2,9)
imshow(mat2gray(reshape(abs(Ur(:,5)),90,430))), colormap('jet');
subplot(9,2,10)
plot(Vrec(:,5));
xlim([1,226])
subplot(9,2,11)
imshow(mat2gray(reshape(abs(Ur(:,6)),90,430))), colormap('jet');
subplot(9,2,12)
plot(Vrec(:,6));
xlim([1,226])
subplot(9,2,13)
imshow(mat2gray(reshape(abs(Ur(:,7)),90,430))), colormap('jet');
subplot(9,2,14)
plot(Vrec(:,7));
xlim([1,226])
subplot(9,2,15)
imshow(mat2gray(reshape(abs(Ur(:,8)),90,430))), colormap('jet');
subplot(9,2,16)
plot(Vrec(:,8));
xlim([1,226])
subplot(9,2,17)
imshow(mat2gray(reshape(abs(Ur(:,9)),90,430))), colormap('jet');
subplot(9,2,18)
plot(Vrec(:,9));
xlim([1,226])

%% Reconstruct flow
Vrecf = Vrec(:,1:30);

filter = fft(Vrecf(:,1));
    filter = fftshift(filter);
    filter(1:80)=0;filter(150:end)=0;
    filter = ifftshift(filter);
    Vrecf(:,1) = real(ifft(filter));

for i = 2:30
    filter = fft(Vrecf(:,i));
    filter = fftshift(filter);
    filter(1:85)=0;filter(135:end)=0;
    filter = ifftshift(filter);
    Vrecf(:,i) = real(ifft(filter));
end


figure('Name','Principal Flow Modes downsampled')
subplot(9,2,1)
imshow(mat2gray(reshape(abs(Ur(:,1)),90,430))), colormap('jet');
title('POD Modes downsampled')
subplot(9,2,2)
title('Corellation to Time reconstructed')
hold on
plot(Vrecf(:,1));
xlim([1,226])
subplot(10,2,3)
imshow(mat2gray(reshape(abs(Ur(:,2)),90,430))), colormap('jet');
subplot(10,2,4)
plot(Vrecf(:,2));
xlim([1,226])
subplot(10,2,5)
imshow(mat2gray(reshape(abs(Ur(:,3)),90,430))), colormap('jet');
subplot(10,2,6)
plot(Vrecf(:,3));
xlim([1,226])
subplot(10,2,7)
imshow(mat2gray(reshape(abs(Ur(:,4)),90,430))), colormap('jet');
subplot(10,2,8)
plot(Vrecf(:,4));
xlim([1,226])
subplot(10,2,9)
imshow(mat2gray(reshape(abs(Ur(:,5)),90,430))), colormap('jet');
subplot(10,2,10)
plot(Vrecf(:,5));
xlim([1,226])
subplot(10,2,11)
imshow(mat2gray(reshape(abs(Ur(:,6)),90,430))), colormap('jet');
subplot(10,2,12)
plot(Vrecf(:,6));
xlim([1,226])
subplot(10,2,13)
imshow(mat2gray(reshape(abs(Ur(:,7)),90,430))), colormap('jet');
subplot(10,2,14)
plot(Vrecf(:,7));
xlim([1,226])
subplot(10,2,15)
imshow(mat2gray(reshape(abs(Ur(:,8)),90,430))), colormap('jet');
subplot(10,2,16)
plot(Vrecf(:,8));
xlim([1,226])
subplot(10,2,17)
imshow(mat2gray(reshape(abs(Ur(:,9)),90,430))), colormap('jet');
subplot(10,2,18)
plot(Vrecf(:,9));
xlim([1,226])
subplot(10,2,19)
imshow(mat2gray(reshape(abs(Ur(:,10)),90,430))), colormap('jet');
subplot(10,2,20)
plot(Vrecf(:,10));
xlim([1,226])

Rec = Ur(:,1:15)*Sr(1:15,1:15)*Vrecf(:,1:15)';

meanr = mean(Rec);
meano = mean(X);
factor = meano/meanr;

Rec = Rec*factor;

for i = 1:226
    img = reshape(Rec(:,i),90,430);
    gif2(:,:,:,i) = mat2gray(img); 
end
implay(gif2);

File: Code Python/07. Physical Models ROMs/07.1 POD on Fluid Flow past a Cylinder/CylinderFlowModes.py
from PIL import Image, ImageSequence
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation
import cvxpy as cvx

# Import Data
gif = Image.open("CylinderFlow.gif")

X = np.empty((38700, 226))
k = 0
for frame in ImageSequence.Iterator(gif):
    x = np.array(frame)
    x = x[90:180, 50:]
    X[:, k] = x.reshape(-1)
    k += 1

fig = plt.figure()
gifup = []
for i in range(gif.n_frames):
    gifup.append([plt.imshow(X[:, i].reshape(90, 430), animated=True, cmap='gray')])

ani = animation.ArtistAnimation(fig, gifup, interval=260, blit=True, repeat_delay=100)
ani.save("Flow.gif")

# Principal Modes
# PCA of timeframes

u, s, v = np.linalg.svd(X, full_matrices=False)

plt.figure()
plt.subplot(1, 2, 1)
plt.title('Singular values')
plt.semilogy(s[:50], 'o-')
plt.subplot(1, 2, 2)
svar = np.empty(s.shape)
for i in range(len(s)):
    svar[i] = sum(s[0:i]) / sum(s)
plt.plot(svar)
plt.title('Captured variance')
plt.show()

# PLot first POD Modes
plt.figure(figsize=(10, 20))
plt.subplot(5, 2, 1)
plt.imshow(np.abs(u[:, 0].reshape(90, 430)), cmap='jet')
plt.title('Mean Flow')
for i in range(3, 11):
    plt.subplot(5, 2, i)
    plt.imshow(np.abs(u[:, i - 2].reshape(90, 430)), cmap='jet')
    plt.title('POD' + str(i - 1))
plt.show()

# Time Correlation

plt.figure(figsize=(10, 20))
for i in range(1, 19, 2):
    plt.subplot(9, 2, i)
    plt.imshow(np.abs(u[:, i - 1].reshape(90, 430)), cmap='jet')
    plt.subplot(9, 2, i + 1)
    plt.plot(v[i - 1, :])
plt.show()

# Compressed Sensing with time samples

R = np.round(np.linspace(0, 216, 40))
R = R + np.random.randint(5, size=40)

Xr = X[:, np.int_(R)]

ur, sr, vr = np.linalg.svd(Xr, full_matrices=False)

plt.figure(figsize=(10, 20))
plt.subplot(5, 2, 1)
plt.imshow(np.abs(ur[:, 0].reshape(90, 430)), cmap='jet')
plt.title('Mean Flow')
for i in range(3, 11):
    plt.subplot(5, 2, i)
    plt.imshow(np.abs(ur[:, i - 2].reshape(90, 430)), cmap='jet')
    plt.title('POD' + str(i - 1))
plt.show()

plt.figure(figsize=(10, 20))
for i in range(1, 19, 2):
    plt.subplot(9, 2, i)
    plt.imshow(np.abs(ur[:, i - 1].reshape(90, 430)), cmap='jet')
    plt.subplot(9, 2, i + 1)
    plt.plot(vr[i - 1, :])
plt.show()

PHI = np.fft.fft(np.eye(226))
PSI = PHI[np.int_(R),:]

vrec = np.empty((v.shape[1], 30))

for i in range(30):
    b = vr[i, :]
    xL1 = cvx.Variable(PSI.shape[1], complex=True)
    objective = cvx.Minimize(cvx.norm(xL1, 1))
    constraints = [cvx.norm(PSI @ xL1 - b, 2) <= .05]
    prob = cvx.Problem(objective, constraints)
    result = prob.solve()
    vrec[:, i] = np.real(np.fft.ifft(xL1.value))

plt.figure(figsize=(10, 20))
for i in range(1, 19, 2):
    plt.subplot(9, 2, i)
    plt.imshow(np.abs(ur[:, i - 1].reshape(90, 430)), cmap='jet')
    plt.subplot(9, 2, i + 1)
    plt.plot(vrec[:, i-1])
plt.show()


# Reconstruct Flow
vrec_f = np.empty(vrec.shape)

f = np.fft.fftshift(np.fft.fft(vrec[:,0]))
f[1:80] = 0
f[150:] = 0
vrec_f[:, 0] = np.real(np.fft.ifft(np.fft.fftshift(f)))

for i in range(1, 30):
    f = np.fft.fftshift(np.fft.fft(vrec[:, i]))
    f[1:85] = 0
    f[135:] = 0
    vrec_f[:, i] = np.real(np.fft.ifft(np.fft.fftshift(f)))

plt.figure(figsize=(10, 20))
for i in range(1, 19, 2):
    plt.subplot(9, 2, i)
    plt.imshow(np.abs(ur[:, i - 1].reshape(90, 430)), cmap='jet')
    plt.subplot(9, 2, i + 1)
    plt.plot(vrec_f[:, i-1])
plt.show()

Rec = (ur[:,0:30].dot(np.diag(sr)[0:30,0:30])).dot(vrec_f[:,0:30].T)

fig = plt.figure()
gifup = []
for i in reversed(list(range(gif.n_frames))):
    gifup.append([plt.imshow(Rec[:, i].reshape(90, 430), animated=True, cmap='gray')])

ani = animation.ArtistAnimation(fig, gifup, interval=260, blit=True, repeat_delay=100)
ani.save("Rec.gif")


